In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Data Analysis

In [2]:
df = pd.read_csv(r'D:\Parkinson_project\data\pd_speech_features.csv')
df.head()

Unnamed: 0 Unnamed: 1 Baseline Features Unnamed: 3 Unnamed: 4 Unnamed: 5  \
0         id     gender               PPE        DFA       RPDE  numPulses   
1          0          1           0.85247    0.71826    0.57227        240   
2          0          1           0.76686    0.69481    0.53966        234   
3          0          1           0.85083    0.67604    0.58982        232   
4          1          0           0.41121    0.79672    0.59257        178   

         Unnamed: 6        Unnamed: 7          Unnamed: 8    Unnamed: 9  ...  \
0  numPeriodsPulses  meanPeriodPulses  stdDevPeriodPulses  locPctJitter  ...   
1               239        0.00806353            8.68E-05       0.00218  ...   
2               233       0.008258256            7.31E-05       0.00195  ...   
3               231        0.00833959            6.04E-05       0.00176  ...   
4               177       0.010857733         0.000182739       0.00419  ...   

                Unnamed: 745               Unnamed: 746  \
0  tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29   
1                      1.562                     2.6445   
2                     1.5589                     3.6107   
3                     1.5643                     2.3308   
4                     3.7805                     3.5664   

                Unnamed: 747               Unnamed: 748  \
0  tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31   
1                     3.8686                     4.2105   
2                    23.5155                    14.1962   
3                     9.4959                    10.7458   
4                     5.2558                    14.0403   

                Unnamed: 749               Unnamed: 750  \
0  tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33   
1                     5.1221                     4.4625   
2                    11.0261                     9.5082   
3                    11.0177                     4.8066   
4                     4.2235                     4.6857   

                Unnamed: 751               Unnamed: 752  \
0  tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35   
1                     2.6202                     3.0004   
2                     6.5245                     6.3431   
3                     2.9199                     3.1495   
4                      4.846                      6.265   

                Unnamed: 753 Unnamed: 754  
0  tqwt_kurtosisValue_dec_36        class  
1                    18.9405            1  
2                     45.178            1  
3                     4.7666            1  
4                     4.0603            1  

[5 rows x 755 columns]

In [3]:
df =df.rename(columns=df.iloc[0])
df = df[1:]
df.head()

id gender      PPE      DFA     RPDE numPulses numPeriodsPulses  \
1  0      1  0.85247  0.71826  0.57227       240              239   
2  0      1  0.76686  0.69481  0.53966       234              233   
3  0      1  0.85083  0.67604  0.58982       232              231   
4  1      0  0.41121  0.79672  0.59257       178              177   
5  1      0   0.3279  0.79782  0.53028       236              235   

  meanPeriodPulses stdDevPeriodPulses locPctJitter  ...  \
1       0.00806353           8.68E-05      0.00218  ...   
2      0.008258256           7.31E-05      0.00195  ...   
3       0.00833959           6.04E-05      0.00176  ...   
4      0.010857733        0.000182739      0.00419  ...   
5      0.008161574        0.002668863      0.00535  ...   

  tqwt_kurtosisValue_dec_28 tqwt_kurtosisValue_dec_29  \
1                     1.562                    2.6445   
2                    1.5589                    3.6107   
3                    1.5643                    2.3308   
4                    3.7805                    3.5664   
5                    6.1727                    5.8416   

  tqwt_kurtosisValue_dec_30 tqwt_kurtosisValue_dec_31  \
1                    3.8686                    4.2105   
2                   23.5155                   14.1962   
3                    9.4959                   10.7458   
4                    5.2558                   14.0403   
5                    6.0805                    5.7621   

  tqwt_kurtosisValue_dec_32 tqwt_kurtosisValue_dec_33  \
1                    5.1221                    4.4625   
2                   11.0261                    9.5082   
3                   11.0177                    4.8066   
4                    4.2235                    4.6857   
5                    7.7817                   11.6891   

  tqwt_kurtosisValue_dec_34 tqwt_kurtosisValue_dec_35  \
1                    2.6202                    3.0004   
2                    6.5245                    6.3431   
3                    2.9199                    3.1495   
4                     4.846                     6.265   
5                    8.2103                    5.0559   

  tqwt_kurtosisValue_dec_36 class  
1                   18.9405     1  
2                    45.178     1  
3                    4.7666     1  
4                    4.0603     1  
5                    6.1164     1  

[5 rows x 755 columns]

In [4]:
df.shape

(756, 755)

In [5]:
df.gender.value_counts()

1    390
0    366
Name: gender, dtype: int64

### Data description:
The dataset that has been deployed for this investigativestudy is taken from the UCI machine learning repository[28]. The Parkinson’s disease (PD) classification dataset isconstituted at the Department of Neurology in CerrahpaÅŸaFaculty of Medicine, Istanbul University .

In total, 188 PDpatients ranging from 33 to 87 were studied where therewere 107 men and 81 women. However, the control group ismade up of 64 healthy individuals ranging in age from 41 to82 years old (23 men and 41 women). There are 756instances and 754 features in the dataset where the features

To extract clinically usefulinformation for PD assessment, various speech signalprocessing algorithms such as Time Frequency Features,Mel Frequency Cepstral Coefficients (MFCCs), WaveletTransform based Features, Vocal Fold Features, and TWQTfeatures were implemented to the speech recordings ofParkinson's Disease (PD) patients. The microphone wasadjusted to 44.1 KHz during the data collection process, andeach subject's sustained phonation of the vowel ‘a’ wastaken three times after the physician's examination. 



In [6]:
df.isnull().sum().any()

False

here were not missing values in the dataset.

In [7]:
df.describe()

id gender      PPE      DFA     RPDE numPulses numPeriodsPulses  \
count   756    756      756      756      756       756              756   
unique  252      2      740      745      748       315              319   
top       0      1  0.82273  0.72248  0.62128       237              236   
freq      3    390        3        2        2         9                8   

       meanPeriodPulses stdDevPeriodPulses locPctJitter  ...  \
count               756                756          756  ...   
unique              755                646          358  ...   
top         0.006004477           7.17E-05      0.00076  ...   
freq                  2                  3            9  ...   

       tqwt_kurtosisValue_dec_28 tqwt_kurtosisValue_dec_29  \
count                        756                       756   
unique                       749                       755   
top                       1.7726                    4.0251   
freq                           2                         2   

       tqwt_kurtosisValue_dec_30 tqwt_kurtosisValue_dec_31  \
count                        756                       756   
unique                       752                       753   
top                       2.8454                    4.2105   
freq                           2                         2   

       tqwt_kurtosisValue_dec_32 tqwt_kurtosisValue_dec_33  \
count                        756                       756   
unique                       749                       752   
top                       2.6829                    3.1144   
freq                           2                         2   

       tqwt_kurtosisValue_dec_34 tqwt_kurtosisValue_dec_35  \
count                        756                       756   
unique                       753                       753   
top                       12.595                    4.2391   
freq                           2                         2   

       tqwt_kurtosisValue_dec_36 class  
count                        756   756  
unique                       754     2  
top                       3.2941     1  
freq                           2   564  

[4 rows x 755 columns]

In [8]:
df['class'].unique()# therefore class is the predictive feature

array(['1', '0'], dtype=object)

from the reference paper, to improve the model performance we can reduce the dimension of the data sets by reducing the feature columns by selecting the best feature 

**highly crrelated features has been found from correlation matrix where the thresholdvalue has been set to 0.75 and thus, the features have beenreduced to 257 from 755**

In [9]:
df = df.drop('id',axis=1)

In [10]:
df.info() # all the features are in object so change them into float

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 1 to 756
Columns: 754 entries, gender to class
dtypes: object(754)
memory usage: 4.3+ MB


In [11]:
df = df.astype('float64')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 1 to 756
Columns: 754 entries, gender to class
dtypes: float64(754)
memory usage: 4.3 MB


In [12]:
# Create correlation matrix
corr_matrix = df.corr().abs()

# Select upper triangle of correlation matrix
upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))

# Find index of feature columns with correlation greater than 0.95
to_drop = [column for column in upper.columns if any(upper[column] > 0.75)]


C:\Users\yoke pc\AppData\Local\Temp\ipykernel_24588\247374459.py:5: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  upper = corr_matrix.where(np.triu(np.ones(corr_matrix.shape), k=1).astype(np.bool))


In [13]:
len(df.columns) - len(to_drop)

257

#### from the refrence paper 

then, highly crrelated features hasbeen found from correlation matrix where the thresholdvalue has been set to 0.75 and thus, the features have beenreduced to 257 from 755 



In [14]:
pd.set_option('display.max_columns',257)

In [15]:
new_df = df.drop(df[to_drop], axis=1)

In [16]:
new_df.head()

gender      PPE      DFA     RPDE  numPulses  stdDevPeriodPulses  \
1     1.0  0.85247  0.71826  0.57227      240.0            0.000087   
2     1.0  0.76686  0.69481  0.53966      234.0            0.000073   
3     1.0  0.85083  0.67604  0.58982      232.0            0.000060   
4     0.0  0.41121  0.79672  0.59257      178.0            0.000183   
5     0.0  0.32790  0.79782  0.53028      236.0            0.002669   

   locPctJitter  locShimmer  meanAutoCorrHarmonicity  minIntensity  \
1       0.00218     0.05883                 0.970805     69.997496   
2       0.00195     0.05516                 0.984322     67.415903   
3       0.00176     0.09902                 0.974846     62.661706   
4       0.00419     0.05451                 0.968343     76.306989   
5       0.00535     0.05610                 0.975754     76.645686   

           f1           f2           f3           f4          b1          b2  \
1  539.342735  1031.849040  2447.162183  3655.054806  101.092218   83.147440   
2  564.363615  1016.367294  2383.565201  3498.681572   58.465428   86.487292   
3  548.444604  1032.406341  2357.826954  3678.128717  160.387771   54.685168   
4  819.529588  1201.813897  3154.035654  4122.163933  238.667052  191.984916   
5  846.796144  1215.346469  3201.513132  4085.456839  402.216738  210.061394   

           b3          b4  GQ_prc5_95  GQ_std_cycle_open  GNE_mean  \
1  255.214830  396.643631     0.77778            11.7245   1.17300   
2  248.357127  218.229722     0.81250            13.8284   1.02210   
3  151.694847   84.240339     0.81818            26.9273   0.84951   
4  573.752909  526.147599     0.98548           139.5744   0.83405   
5  203.637106  384.611697     0.97847           102.0549   0.83556   

   GNE_SNR_TKEO  GNE_SNR_SEO  GNE_NSR_SEO  VFER_mean  VFER_std  VFER_SNR_TKEO  \
1      0.083127  1200445.612       3.0152   0.000463  0.000260       209.6062   
2      0.127410  1298455.445       3.0600   0.000615  0.000295       243.0816   
3      0.116890  1272869.841       3.0309   0.000360  0.000266       238.5976   
4      0.147370  1932289.206       3.0293   0.000196  0.000194        79.0921   
5      0.151990  1861807.802       3.0362   0.000281  0.000175        53.6764   

   VFER_SNR_SEO  VFER_NSR_TKEO  VFER_NSR_SEO  IMF_SNR_SEO  IMF_SNR_TKEO  \
1      455.9654         1.2825        1.3305      51.6843        5.7784   
2      379.8429         1.3063        1.3177      24.0230        6.7916   
3      386.4739         1.2614        1.3078      60.0458       11.3376   
4      161.1054         1.2369        1.2986      83.6201        5.0784   
5      164.4029         1.2360        1.3156     102.9371        5.6022   

   IMF_NSR_SEO  IMF_NSR_TKEO  mean_Log_energy  mean_MFCC_0th_coef  \
1      0.26850        5.8573           9.1817             11.4283   
2      0.26839        6.2366           9.6074             12.1387   
3      0.23034        4.8081           9.0512             11.3956   
4      0.23529        4.2629           9.8453             13.7399   
5      0.30603        6.3500          10.1542             14.7643   

   mean_MFCC_1st_coef  mean_MFCC_2nd_coef  mean_MFCC_3rd_coef  \
1              8.4781              2.4874             0.02292   
2              8.5008              2.8986            -0.46733   
3              7.6362              3.2208            -0.48228   
4              6.7365              3.1023             0.75419   
5              6.4634              2.9451             0.83210   

   mean_MFCC_4th_coef  mean_MFCC_5th_coef  mean_MFCC_6th_coef  \
1             0.55527             -1.6334            -1.51440   
2             0.45984             -1.3427            -1.67590   
3             0.28918             -1.7334            -1.49290   
4            -1.55900             -1.0083            -0.65556   
5            -1.75500             -1.0392            -0.51876   

   mean_MFCC_7th_coef  mean_MFCC_8th_coef  mean_MFCC_9th_coef  \
1             0.39725             0.61775            

Finally, top twenty features werecalculated utilizing Kbest algorithm that takes a scoringfunction as a parameter, which must be relevant to a pair(X, y). The score function must return an array of scores,one for each feature and then simply holds thefirst ‘k’ features of ‘X’ with the highest scores 


In [21]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

y = df['class']
X = df.drop('class',axis=1)
X = X.abs()
X

chi2_features = SelectKBest(chi2,k = 20)
new_df = chi2_features.fit_transform(X, y)


In [22]:
new_df = pd.DataFrame(new_df)
new_df.head()

0            1            2            3             4   \
0  1.652570e+02  34415952.86  41505780.29  54991273.00  7.977922e+07   
1  6.501035e+02  32588713.93  39329560.14  52167450.85  7.573630e+07   
2  1.595700e+00  31864278.37  38418848.45  50880181.14  7.373527e+07   
3  4.520939e+06  44854174.87  50479679.82  60032621.59  7.660933e+07   
4  2.797297e+06  59737876.86  66995068.86  80313780.59  1.003577e+08   

            5            6            7            8             9   \
0  129684181.7  232253303.7  459058057.8  902118385.7  1.886559e+09   
1  123243056.6  220947833.1  437011684.1  859038700.4  1.796706e+09   
2  119780270.0  214375881.7  423881698.8  832850139.0  1.742079e+09   
3  109696232.2  176812107.4  330129338.5  625193831.3  1.306602e+09   
4  140671395.9  225887933.1  425927317.4  786450935.2  1.640684e+09   

             10           11           12            13           14  \
0  3.937753e+09  245931.5887  524800.1114  1.049747e+06  2101636.245   
1  3.750669e+09  236033.1031  502899.2766  1.005893e+06  2012043.365   
2  3.636913e+09  230020.8923  490695.5293  9.813527e+05  1962144.341   
3  2.722133e+09  184765.6410  377576.4933  7.358128e+05  1507172.433   
4  3.391442e+09  404982.5071  853765.7904  1.713112e+06  3158402.899   

            15            16           17           18           19  
0  326874.1892  6.720895e+05  1381934.183  2777808.879  5547195.189  
1  314467.5684  6.427345e+05  1327168.357  2654568.330  5309694.850  
2  305549.8724  6.285763e+05  1295706.406  2588074.975  5177908.600  
3  314057.2871  7.030022e+05  1002359.597  2359863.040  4011147.820  
4  713080.2915  1.119327e+06  2556027.478  4504909.124  8579124.177

In [ ]:
# there the reduced features are from the kbest algo we got the best 20 features
columns = ['numPulses','GNE_NSR_SEO','mean_MFCC_1st_coef','mean_MFCC_6th_coef','std_MFCC_8th_coef','tqwt_energy_dec_27','tqwt_energy_dec_31','tqwt_energy_dec_35','tqwt_entropy_shannon_dec_32','tqwt_TKEO_mean_dec_21','tqwt_medianValue_dec_25','tqwt_medianValue_dec_26','tqwt_medianValue_dec_34','tqwt_meanValue_dec_9','tqwt_meanValue_dec_11','tqwt_meanValue_dec_21','tqwt_meanValue_dec_26','tqwt_skewnessValue_dec_3','tqwt_skewnessValue_dec_21','tqwt_skewnessValue_dec_28','class']
reduced_df = df[columns]
reduced_df.head()

numPulses  GNE_NSR_SEO  mean_MFCC_1st_coef  mean_MFCC_6th_coef  \
1      240.0       3.0152              8.4781            -1.51440   
2      234.0       3.0600              8.5008            -1.67590   
3      232.0       3.0309              7.6362            -1.49290   
4      178.0       3.0293              6.7365            -0.65556   
5      236.0       3.0362              6.4634            -0.51876   

   std_MFCC_8th_coef  tqwt_energy_dec_27  tqwt_energy_dec_31  \
1            0.17101            0.108070            0.000007   
2            0.16785            0.098360            0.000015   
3            0.21901            0.106910            0.000026   
4            0.24529            0.002072            0.000207   
5            0.22090            0.001152            0.000166   

   tqwt_energy_dec_35  tqwt_entropy_shannon_dec_32  tqwt_TKEO_mean_dec_21  \
1            0.000001                     0.036109               0.010913   
2            0.000002                     0.068996               0.007092   
3            0.000004                     0.050766               0.002236   
4            0.000054                     4.267800               0.105100   
5            0.000028                     3.941400               0.118100   

   tqwt_medianValue_dec_25  tqwt_medianValue_dec_26  tqwt_medianValue_dec_34  \
1                 0.000000            -2.220000e-17                -0.000202   
2                 0.000000            -3.790000e-17                -0.000158   
3                 0.000000            -1.580000e-17                -0.000136   
4                 0.002295             1.105400e-03                -0.000313   
5                 0.017662             8.909500e-03                 0.000663   

   tqwt_meanValue_dec_9  tqwt_meanValue_dec_11  tqwt_meanValue_dec_21  \
1          1.230000e-20          -1.630000e-20          -6.330000e-20   
2          2.030000e-20          -1.680000e-20           3.090000e-19   
3         -1.050000e-21           7.080000e-22           1.760000e-19   
4          2.940000e-20           5.630000e-20           7.780000e-19   
5         -1.170000e-20           1.280000e-20          -1.700000e-18   

   tqwt_meanValue_dec_26  tqwt_skewnessValue_dec_3  tqwt_skewnessValue_dec_21  \
1          -7.210000e-19              1.110000e-15              -4.470000e-17   
2           3.150000e-19             -3.290000e-15              -7.020000e-20   
3          -2.700000e-19             -3.420000e-15               1.200000e-16   
4          -2.340000e-18              6.455000e-02              -1.281100e-03   
5          -8.110000e-19             -4.187600e-03               3.441400e-03   

   tqwt_skewnessValue_dec_28  class  
1                  -0.102900    1.0  
2                  -0.005679    1.0  
3                  -0.003824    1.0  
4                   0.027619    1.0  
5                   0.011756    1.0

In [ ]:
reduced_df.to_csv(r"D:\Parkinson_project\data\featured_data.csv")

In [ ]:
reduced_df.shape

(756, 21)